// mindee/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductConfigs = void 0;
const api_1 = require("./api");
const documents_1 = require("./documents");
class ProductConfigs {
    static getByResponseClassName(responseClass) {
        const config = this.configs.find((element) => element.responseClass.name === responseClass);
        if (config === undefined) {
            throw `Unknown response class: ${responseClass}`;
        }
        return config;
    }
    static getByDocType(docType) {
        const config = this.configs.find((element) => element.docType === docType);
        if (config === undefined) {
            throw `Unknown document type: ${docType}`;
        }
        return config;
    }
}
exports.ProductConfigs = ProductConfigs;
ProductConfigs.configs = [
    {
        description: "Invoice",
        docType: documents_1.DOC_TYPE_INVOICE,
        responseClass: api_1.InvoiceResponse,
        fullText: true,
    },
    {
        description: "Expense Receipt",
        docType: documents_1.DOC_TYPE_RECEIPT,
        responseClass: api_1.ReceiptResponse,
        fullText: true,
    },
    {
        description: "Passport",
        docType: documents_1.DOC_TYPE_PASSPORT,
        responseClass: api_1.PassportResponse,
        fullText: false,
    },
    {
        description: "Financial Document (receipt or invoice)",
        docType: documents_1.DOC_TYPE_FINANCIAL,
        responseClass: api_1.FinancialDocResponse,
        fullText: true,
    },
    {
        description: "A custom document",
        docType: documents_1.DOC_TYPE_CUSTOM,
        responseClass: api_1.CustomResponse,
        fullText: false,
    },
    {
        description: "Cropper",
        docType: documents_1.DOC_TYPE_CROPPER,
        responseClass: api_1.CropperResponse,
        fullText: false,
    },
];
