// mindee/src/documents/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DOC_TYPE_CROPPER = exports.DOC_TYPE_CUSTOM = exports.DOC_TYPE_FINANCIAL = exports.DOC_TYPE_PASSPORT = exports.DOC_TYPE_RECEIPT = exports.DOC_TYPE_INVOICE = exports.Cropper = exports.CustomDocument = exports.Passport = exports.FinancialDocument = exports.Invoice = exports.Receipt = exports.Document = void 0;
const invoice_1 = require("./invoice");
Object.defineProperty(exports, "Invoice", { enumerable: true, get: function () { return invoice_1.Invoice; } });
const receipt_1 = require("./receipt");
Object.defineProperty(exports, "Receipt", { enumerable: true, get: function () { return receipt_1.Receipt; } });
const passport_1 = require("./passport");
Object.defineProperty(exports, "Passport", { enumerable: true, get: function () { return passport_1.Passport; } });
const financialDocument_1 = require("./financialDocument");
Object.defineProperty(exports, "FinancialDocument", { enumerable: true, get: function () { return financialDocument_1.FinancialDocument; } });
const custom_1 = require("./custom");
Object.defineProperty(exports, "CustomDocument", { enumerable: true, get: function () { return custom_1.CustomDocument; } });
const cropper_1 = require("./cropper");
Object.defineProperty(exports, "Cropper", { enumerable: true, get: function () { return cropper_1.Cropper; } });
var document_1 = require("./document");
Object.defineProperty(exports, "Document", { enumerable: true, get: function () { return document_1.Document; } });
exports.DOC_TYPE_INVOICE = invoice_1.Invoice.name;
exports.DOC_TYPE_RECEIPT = receipt_1.Receipt.name;
exports.DOC_TYPE_PASSPORT = passport_1.Passport.name;
exports.DOC_TYPE_FINANCIAL = financialDocument_1.FinancialDocument.name;
exports.DOC_TYPE_CUSTOM = custom_1.CustomDocument.name;
exports.DOC_TYPE_CROPPER = cropper_1.Cropper.name;
