// mrz/lib-esm/parse/getResult.js
'use strict';
function getDetails(lines, fieldParsers, autocorrectArray) {
    const details = [];
    fieldParsers.forEach(({ parser }, i) => {
        details.push(parser(lines, autocorrectArray[i]));
    });
    return details;
}
function getFields(details) {
    const fields = {};
    let valid = true;
    for (const detail of details) {
        if (!detail.valid)
            valid = false;
        if (detail.field) {
            fields[detail.field] = detail.value;
        }
    }
    return { fields, valid };
}
function getCorrection(lines, fieldParsers, autocorrect) {
    let corrected = lines;
    const autocorrectArray = [];
    if (autocorrect) {
        fieldParsers.forEach(({ autocorrector }) => {
            const result = autocorrector(corrected);
            autocorrectArray.push(result.autocorrect);
            corrected = result.correctedLines;
        });
    }
    return { corrected, autocorrectArray };
}
export function getResult(format, lines, fieldParsers, options) {
    const { autocorrect = false } = options;
    const { corrected, autocorrectArray } = getCorrection(lines, fieldParsers, autocorrect);
    const details = getDetails(corrected, fieldParsers, autocorrectArray);
    const fields = getFields(details);
    const result = {
        format,
        details,
        fields: fields.fields,
        valid: fields.valid,
    };
    return result;
}
