// mindee/src/api/endpoint.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomEndpoint = exports.CropperEndpoint = exports.PassportEndpoint = exports.ReceiptEndpoint = exports.InvoiceEndpoint = exports.Endpoint = exports.API_KEY_ENVVAR_NAME = exports.STANDARD_API_OWNER = void 0;
const https = __importStar(require("https"));
const os = __importStar(require("os"));
const package_json_1 = require("../../package.json");
const url_1 = require("url");
const form_data_1 = __importDefault(require("form-data"));
const logger_1 = require("../logger");
const MINDEE_API_URL = "https://api.mindee.net/v1";
const USER_AGENT = `mindee-api-nodejs@v${package_json_1.version} nodejs-${process.version} ${os.type().toLowerCase()}`;
exports.STANDARD_API_OWNER = "mindee";
exports.API_KEY_ENVVAR_NAME = "MINDEE_API_KEY";
class Endpoint {
    constructor(owner, urlName, version, apiKey, keyName) {
        this.owner = owner;
        this.urlName = urlName;
        this.version = version;
        this.keyName = keyName || urlName;
        this.apiKey = apiKey || this.apiKeyFromEnv();
        this.urlRoot = `${MINDEE_API_URL}/products/${owner}/${urlName}/v${version}`;
    }
    /**
     * Make a prediction request.
     */
    predictRequest(input, includeWords = false, cropper = false) {
        return new Promise((resolve, reject) => {
            let headers = {
                "User-Agent": USER_AGENT,
                Authorization: `Token ${this.apiKey}`,
            };
            const uri = new url_1.URL(`${this.urlRoot}/predict`);
            if (cropper) {
                uri.searchParams.append("cropper", "true");
            }
            logger_1.logger.debug(`Prediction request: ${uri}`);
            const form = new form_data_1.default();
            const fileParams = { filename: input.filename };
            form.append("document", input.fileObject, fileParams);
            if (includeWords) {
                form.append("include_mvision", "true");
            }
            headers = { ...headers, ...form.getHeaders() };
            const options = {
                method: "POST",
                headers: headers,
                hostname: uri.hostname,
                path: `${uri.pathname}${uri.search}`,
            };
            const req = https.request(options, function (res) {
                // when the encoding is set, data chunks will be strings
                res.setEncoding("utf-8");
                let responseBody = "";
                res.on("data", function (chunk) {
                    responseBody += chunk;
                });
                res.on("end", function () {
                    try {
                        resolve({
                            ...res,
                            data: JSON.parse(responseBody),
                        });
                    }
                    catch (error) {
                        reject(error);
                    }
                });
            });
            form.pipe(req);
            req.on("error", (err) => {
                reject(err);
            });
        });
    }
    apiKeyFromEnv() {
        const envVarValue = process.env[exports.API_KEY_ENVVAR_NAME];
        if (envVarValue) {
            logger_1.logger.debug(`Set '${this.keyName}' API key from environment: ${exports.API_KEY_ENVVAR_NAME}`);
            return envVarValue;
        }
        return "";
    }
}
exports.Endpoint = Endpoint;
class InvoiceEndpoint extends Endpoint {
    constructor(apiKey) {
        super(exports.STANDARD_API_OWNER, "invoices", "3", apiKey, "invoice");
    }
}
exports.InvoiceEndpoint = InvoiceEndpoint;
class ReceiptEndpoint extends Endpoint {
    constructor(apiKey) {
        super(exports.STANDARD_API_OWNER, "expense_receipts", "3", apiKey, "receipt");
    }
}
exports.ReceiptEndpoint = ReceiptEndpoint;
class PassportEndpoint extends Endpoint {
    constructor(apiKey) {
        super(exports.STANDARD_API_OWNER, "passport", "1", apiKey);
    }
}
exports.PassportEndpoint = PassportEndpoint;
class CropperEndpoint extends Endpoint {
    constructor(apiKey) {
        super(exports.STANDARD_API_OWNER, "cropper", "1", apiKey);
    }
}
exports.CropperEndpoint = CropperEndpoint;
class CustomEndpoint extends Endpoint {
    constructor(documentType, accountName, version, apiKey) {
        super(accountName, documentType, version, apiKey);
    }
}
exports.CustomEndpoint = CustomEndpoint;
